File: tests/bridges/scripts/t_config.py
"""Reads its config in the constructor and reports it, so the whole channel is exercised from a script.

The constructor is where self.config has to be readable: that is the point of a channel that arrives
before initialize, and reading it here is what proves the binding order.
"""

import atp


class Configured(atp.Module):
    name = "py_config"

    out_report = atp.Output(atp.text)

    def __init__(self):
        cfg = self.config
        if cfg is None:
            self.report = "config=None"
            return
        self.report = "channels={} name={} nested={} keys={}".format(
            cfg["channels"][2],
            cfg["name"],
            cfg["nested"]["deep"],
            ",".join(cfg.keys()),
        )

    def iterate(self):
        if self.report is None:
            return atp.IDLE
        self.out_report.write(self.report)
        self.report = None
        return atp.BUSY

File: tests/bridges/lua_scripts/t_config.lua
local atp = require("atp")

local M = atp.module("lua_config", { 1, 0 })

M.out_report = atp.output(atp.text)

function M:initialize()
    if self.config == nil then
        self.report = "config=nil"
        return
    end
    local keys = {}
    for key in pairs(self.config) do
        keys[#keys + 1] = key
    end
    table.sort(keys)
    self.report = string.format("channels=%d name=%s nested=%s keys=%s", self.config.channels[3], self.config.name,
        tostring(self.config.nested.deep), table.concat(keys, ","))
end

function M:iterate()
    if self.report == nil then
        return atp.IDLE
    end
    self.out_report:write(self.report)
    self.report = nil
    return atp.BUSY
end

return M
